core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java

```diff
@@ -306,26 +307,23 @@ void build(List<StaticBytecodeRecorderBuildItem> staticInitTasks,
         CatchBlockCreator preventFurtherStepsBlock = tryBlock.addCatch(PreventFurtherStepsException.class);
         preventFurtherStepsBlock.invokeVirtualMethod(ofMethod(StartupContext.class, "close", void.class), startupContext);
 
+        CatchBlockCreator catchConfigurationException = tryBlock.addCatch(ConfigurationException.class);
+
+        // Handle Configuration Exception thrown at startup
+        handleThrownException(launchMode, catchConfigurationException);
+
+        catchConfigurationException.invokeStaticMethod(ofMethod(ApplicationStateNotification.class,
+                "notifyStartupFailed", void.class, Throwable.class),
+                catchConfigurationException.getCaughtException());
+        catchConfigurationException.invokeVirtualMethod(ofMethod(StartupContext.class,
+                "close", void.class), startupContext);
+        catchConfigurationException.throwException(catchConfigurationException.getCaughtException());
```

Comment:
From what I understand, you copied the error handling code from further up where we set up logging.

I'm... not entirely sure this is appropriate. Granted, it's a bit of a mess right now, but from what I can see that code you copied is a bit special, and for reasons I don't quite understand it does things that are normally supposed to be done in `io.quarkus.runtime.Application#start`.

I think it would be safer to avoid this change if we can. Do you think you could just adapt the pre-existing catch block to wrap the exception conditionally? E.g. just replace the `RuntimeException` throwing with something like:

```java
if (e instanceof ConfigurationException) {
    throw new ConfigurationException( "Failed to configure Quarkus", e );
}
else {
    throw new RuntimeException( "Failed to start Quarkus", e );
}
```

Then the change would have a smaller impact and we wouldn't risk breaking something we (I) don't understand.

Reply:
I changed to what You suggested, and now it is throwing exception conditionally, other changes reverted to what was before